<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Tutorial &#8212; pyinseq 0.2.0-beta documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinxcontrib-images/LightBox2/lightbox2/css/lightbox.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2.0-beta',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2/js/lightbox.min.js"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2-customize/jquery-noconflict.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="top" title="pyinseq 0.2.0-beta documentation" href="index.html" />
    <link rel="next" title="Troubleshooting" href="Troubleshooting.html" />
    <link rel="prev" title="About" href="About.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="tutorial">
<span id="tutorial"></span><h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h1>
<div class="section" id="requirements">
<span id="requirements"></span><h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<p>Mac or Linux-based operating systems.</p>
</div>
<div class="section" id="install-python">
<span id="install-python"></span><h2>Install Python<a class="headerlink" href="#install-python" title="Permalink to this headline">¶</a></h2>
<p>Install the <a class="reference external" href="https://www.continuum.io/downloads">Anaconda Python 3.5 download</a>.</p>
</div>
<div class="section" id="install-pyinseq">
<span id="install-pyinseq"></span><h2>Install pyinseq<a class="headerlink" href="#install-pyinseq" title="Permalink to this headline">¶</a></h2>
<p>Install from Github.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">git</span><span class="o">+</span><span class="n">git</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">mandel01</span><span class="o">/</span><span class="n">pyinseq</span>
</pre></div>
</div>
<p>Test for correct installation</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">make</span> <span class="n">test</span>
</pre></div>
</div>
</div>
<div class="section" id="command-line-operation">
<span id="command-line-operation"></span><h2>Command line operation<a class="headerlink" href="#command-line-operation" title="Permalink to this headline">¶</a></h2>
<p>Basic operation and a short description of the files are listed here. Below are detailed desriptions and links to example input files.</p>
<div class="highlight-txt"><div class="highlight"><pre><span></span>pyinseq -i &lt;input file&gt; -s &lt;sample file&gt; -g &lt;genbank file&gt; -e &lt;experiment name&gt;
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">-i</span></code> / <code class="docutils literal"><span class="pre">--input</span></code></p>
<ul class="simple">
<li>Illumina reads file in FASTQ or gzipped-FASTQ format.</li>
</ul>
<p><code class="docutils literal"><span class="pre">-s</span></code> / <code class="docutils literal"><span class="pre">--samples</span></code></p>
<ul class="simple">
<li>Sample list where each line is the <code class="docutils literal"><span class="pre">sample_name</span></code> &lt;tab&gt; <code class="docutils literal"><span class="pre">barcode</span> <span class="pre">(4bp)</span></code>.</li>
</ul>
<p><code class="docutils literal"><span class="pre">-g</span></code> / <code class="docutils literal"><span class="pre">--genome</span></code></p>
<ul class="simple">
<li>Concatenated GenBank File for all contigs for the genome.</li>
</ul>
<p><code class="docutils literal"><span class="pre">-e</span></code> / <code class="docutils literal"><span class="pre">--experiment</span></code></p>
<ul class="simple">
<li>All results files will be created in the subfolder of this name.</li>
</ul>
</div>
<div class="section" id="optional-arguments">
<span id="optional-arguments"></span><h2>Optional arguments<a class="headerlink" href="#optional-arguments" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">-d</span></code> / <code class="docutils literal"><span class="pre">--disruption</span></code></p>
<ul class="simple">
<li>Five-prime fraction of gene (<code class="docutils literal"><span class="pre">0.0</span></code> - <code class="docutils literal"><span class="pre">1.0</span></code>) that must be disrupted for the hit to be counted in the summary_gene_table. Often insertions at the 3&#8217; end of a gene do not disrupt function so it may be of interest to run the pipeline with a disruption value of <code class="docutils literal"><span class="pre">0.8</span></code> or <code class="docutils literal"><span class="pre">0.9</span></code>.</li>
</ul>
</div>
<div class="section" id="output-files">
<span id="output-files"></span><h2>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<div class="section" id="results-directory">
<span id="results-directory"></span><h3><code class="docutils literal"><span class="pre">results/</span></code> directory<a class="headerlink" href="#results-directory" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">summary_gene_table.text</span></code>: summary for entire experiment</li>
<li><code class="docutils literal"><span class="pre">&lt;sample&gt;_sites.txt</span></code> (for each sample): Counts of each insertion in each sample</li>
<li><code class="docutils literal"><span class="pre">&lt;sample&gt;_genes.txt</span></code> (for each sample): Counts of each insertion mapped to genes</li>
<li><code class="docutils literal"><span class="pre">&lt;sample&gt;_bowtie.txt</span></code> (for each sample): Bowtie mapping results</li>
<li><code class="docutils literal"><span class="pre">&lt;sample&gt;_trimmed.fastq</span></code> (for each sample): Demultiplexed fastq reads trimmed for the chromosome sequence only</li>
</ul>
</div>
<div class="section" id="results-genome-lookup-subdirectory">
<span id="results-genome-lookup-subdirectory"></span><h3><code class="docutils literal"><span class="pre">results/genome_lookup/</span></code> subdirectory<a class="headerlink" href="#results-genome-lookup-subdirectory" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">genome.fna</span></code>: genome fasta nucleotide file</li>
<li><code class="docutils literal"><span class="pre">genome.ftt</span></code>: genome feature file</li>
<li>bowtie indexes</li>
</ul>
</div>
<div class="section" id="results-raw-data-subdirectory">
<span id="results-raw-data-subdirectory"></span><h3><code class="docutils literal"><span class="pre">results/raw_data/</span></code> subdirectory<a class="headerlink" href="#results-raw-data-subdirectory" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&lt;sample&gt;.fastq</span></code> (for each sample): demultiplexed files for each sample/barcode</li>
<li><code class="docutils literal"><span class="pre">_other.fastq</span></code>: demultiplexed files for unrecognized barcodes</li>
</ul>
</div>
</div>
<div class="section" id="files-needed">
<span id="files-needed"></span><h2>Files needed<a class="headerlink" href="#files-needed" title="Permalink to this headline">¶</a></h2>
<p><strong>Illumina sequencing reads [<code class="docutils literal"><span class="pre">-i</span></code>]</strong> for the experiment. File can be uncompressed (.fastq or .fq) or gzip-compressed (.fastq.gz or .fq.gz).</p>
<p><a class="reference external" href="../../../pyinseq/tests/data/input/example01.fastq">Example input file</a></p>
<div class="highlight-txt"><div class="highlight"><pre><span></span>@DGL9ZZQ1:720:C6YD0ACXX:2:1101:1246:2185 1:N:0:
GAAGCGACACCGACGACGGTAACAGGTTGGATGATAAGTCCCCGGTCTTCG
+
CCCFFFDDHHHHHCGHHHIJDHIIJJFHHJIJJJJIJJDHIJJHIAGIJJJ
...
</pre></div>
</div>
<p><strong>Sample file [<code class="docutils literal"><span class="pre">-s</span></code>]</strong> describing the sample names and barcodes. Sample names should be restricted to letters, numbers, dash (-), and underscore (_), with a tab between the sample and the barcode in each row of a text file. It is recommended that the file be prepared in a text editor to ensure that additional hidden characters are not introduced. <a class="reference external" href="http://www.barebones.com/products/TextWrangler/">Textwrangler</a> is recommended for Mac.</p>
<p><a class="reference external" href="../../../pyinseq/tests/data/input/example01.txt">Example sample file</a></p>
<div class="highlight-txt"><div class="highlight"><pre><span></span>E001_01 GAAG
E001_02 CTTT
</pre></div>
</div>
<p><strong>GenBank file [-g]</strong> listing the features and DNA sequence for the organism. If the organism has multiple chromosomes/contigs in the sequence the file they should be concatenated into a single file. Ensure that the double slash <code class="docutils literal"><span class="pre">//</span></code> at the end of the file remains to separate each contig.</p>
<p><a class="reference external" href="../../../pyinseq/tests/data/input/ES114v2.gb">Example GenBank file</a></p>
<div class="highlight-txt"><div class="highlight"><pre><span></span>LOCUS       CP000020             2897536 bp    DNA     circular BCT 02-APR-2008
DEFINITION  Vibrio fischeri ES114 chromosome I, complete sequence.
ACCESSION   CP000020
VERSION     CP000020.2  GI:171902228
KEYWORDS    .
SOURCE      Vibrio fischeri ES114
...
CDS             complement(313..747)
                /gene=&quot;mioC&quot;
                /locus_tag=&quot;VF_0001&quot;
...
ORIGIN
        1 aagatcactt aatatatata agatctttta aagagatctt ttattagatc tattatatag
       61 atcgtcgatc tctgtggata agtgataaat gatcaatagg atcatatact ttagatggat
      121 ccaaagttgt tatctttctt tgatcttcga tcggacagct tgaggacaaa agagttagtt
      181 atccacaagg ggggagggcg ttagatctta ttcaatggat aactataact tgatcactgg
      241 atctttctat agttatccac atagtaggta tcatctattt aataactttt atagatcgga
      301 caacactttt tattaacaaa tgtgttgttt tagccacaat tctgctggtt cttcagggat
      361 actattttga gttacatcta tctctaatag agggtacttt tcgatagcgc catgctcttt
      421 taaggtattt tgaagtgatt ttcctgctgc gcagaaagtg tcataacttg aatcaccgat
      481 agcaacaaca gcaaattcaa cgctcgataa tggctgagtc acgctttcta actgttgaat
      541 aaatggttta atgttgtcag ggtaatcacc agcaccgtga gttgatacaa cgagtaacca
      601 taagctatca atatcaatat catctaaatt tggttgatta tgaatatcgg tggaaaaatc
      661 catttcttct aataaatcag caagatggtc accaacatac tcagcaccac ctagagtgct
      721 tcctgtaata atagatactt ttttcatgaa tttatcctat aaaaatataa aaaatgggcc
      781 tacataggcc cattattaat cttattaata ttggttttat ttaccaatac agaatgaagt
...
//
LOCUS       CP000021             1330333 bp    DNA     circular BCT 02-APR-2008
DEFINITION  Vibrio fischeri ES114 chromosome II, complete sequence.
...
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Tutorial</a><ul>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
<li><a class="reference internal" href="#install-python">Install Python</a></li>
<li><a class="reference internal" href="#install-pyinseq">Install pyinseq</a></li>
<li><a class="reference internal" href="#command-line-operation">Command line operation</a></li>
<li><a class="reference internal" href="#optional-arguments">Optional arguments</a></li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#results-directory"><code class="docutils literal"><span class="pre">results/</span></code> directory</a></li>
<li><a class="reference internal" href="#results-genome-lookup-subdirectory"><code class="docutils literal"><span class="pre">results/genome_lookup/</span></code> subdirectory</a></li>
<li><a class="reference internal" href="#results-raw-data-subdirectory"><code class="docutils literal"><span class="pre">results/raw_data/</span></code> subdirectory</a></li>
</ul>
</li>
<li><a class="reference internal" href="#files-needed">Files needed</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="About.html" title="previous chapter">About</a></li>
      <li>Next: <a href="Troubleshooting.html" title="next chapter">Troubleshooting</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, Mark Mandel.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.4.8</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.9</a>
      
    </div>

    

    
  </body>
</html>